src/mito2/src/compaction.rs

```diff
@@ -429,31 +457,211 @@ impl CompactionScheduler {
         };
 
         // Create a local compaction task.
-        let mut local_compaction_task = Box::new(CompactionTaskImpl {
+        let local_compaction_task = Box::new(CompactionTaskImpl {
             request_sender,
             waiters,
             start_time,
             listener,
             picker_output,
             compaction_region,
             compactor: Arc::new(DefaultCompactor {}),
+            memory_guard: None,
         });
 
-        // Submit the compaction task.
+        let estimated_bytes = estimate_compaction_bytes(&local_compaction_task.picker_output);
+        self.try_submit_with_memory(local_compaction_task, region_id, estimated_bytes)
+    }
+
+    fn submit_compaction_task(
+        &mut self,
+        mut task: Box<CompactionTaskImpl>,
+        region_id: RegionId,
+    ) -> Result<()> {
         self.scheduler
             .schedule(Box::pin(async move {
                 INFLIGHT_COMPACTION_COUNT.inc();
-                local_compaction_task.run().await;
+                task.run().await;
                 INFLIGHT_COMPACTION_COUNT.dec();
             }))
             .map_err(|e| {
                 error!(e; "Failed to submit compaction request for region {}", region_id);
-                // If failed to submit the job, we need to remove the region from the scheduler.
                 self.region_status.remove(&region_id);
                 e
             })
     }
 
+    fn try_submit_with_memory(
+        &mut self,
+        mut task: Box<CompactionTaskImpl>,
+        region_id: RegionId,
+        requested_bytes: u64,
+    ) -> Result<()> {
+        match self.memory_manager.try_acquire(requested_bytes) {
+            Some(guard) => {
+                task.memory_guard = Some(guard);
+                self.submit_compaction_task(task, region_id)
+            }
+            None => self.handle_memory_exhausted(region_id, requested_bytes, task),
+        }
```

Comment:
I think we should acquire the permits in the compaction task instead of in the CompactionScheduler. Because the compaction scheduler can schedule many regions, but only max_bg_compactions tasks will execute concurrently. Acquiring permits inside the task is more accurate than inside the CompactionScheduler.

We can also wait inside the task (maybe with a timeout), which won't block the region worker and act as a rate limit mechanism.